src/datasources/data-source.ts

```diff
@@ -0,0 +1,29 @@
+import { openmrsFetch, OpenmrsResource } from '@openmrs/esm-framework';
+import { DataSource } from '../api/types';
+
+export class BaseOpenMRSDataSource implements DataSource<any> {
+  private initialUrl: string;
+  private url: string;
+
+  constructor(url: string) {
+    this.initialUrl = url;
+    this.url = url;
+  }
+
+  fetchData(searchTerm: string, config?: Record<string, any>): Promise<any[]> {
+    this.url = this.initialUrl;
+    if (config?.tag) {
```

Comment:
There seems to be too many things going on here. If `tag` is set we set both the `tag` and the returned value. It seems like it would be better to just have this in an `OpenmrsLocationDataSource` class or something. (We don't actually have a `tag` concept for any other core data type).

Reply:
The datasource module has been implemented in such a way that to define a new data source, you add an entry in the registry with an id and initialize the datasource component with a url. This should be ok for most datasources and in the json you would only need to specify:
`"datasource" : {
    "id": "datasource_id"
}`
However there are datasources that would need extra parameters such as location which has a `tag` parameter. This needs to be passed at the json because the user needs to determine which locations they want for a particular form. This is what led to the introduction of the config parameter passed in the json file

Reply:
That makes sense to me. On the Angular side of things, the data source interface is less strictly-defined, but most data sources implement an interface like:

```
interface DataSource<T> {
  searchOptions: (q: string) => Array<T>;
  resolveSelectedValue: (value: string) => T
}
```

Each data source can support some degree of configuration, but that's effectively passed to the constructor for the data source rather than as an argument to the `fetchData()` function.

Reply:
@ibacher & @samuelmale I have refactored the code to make the base implementation as generic as possible. For special cases like location which may require extra parameters, we introduce a new component.